src/v/cluster/partition.cc

```diff
@@ -537,7 +540,7 @@ partition::timequery(storage::timequery_config cfg) {
             local_query_cfg.min_offset = std::max(
               log()->from_log_offset(_raft->start_offset()),
               local_query_cfg.min_offset);
-            co_return co_await local_timequery(local_query_cfg);
+            co_return co_await local_timequery(local_query_cfg, false);
```

Comment:
>Previous code contained a bug

Just making sure, this was the bug, right? It isn't quite clear from the commit message -- could you elaborate on it there?

Reply:
Note: This was already merged in a different PR.

The problem was that may_answer_from_cloud was set to true even if trim-prefix or retention did truncate the entire cloud partition. `_cloud_storage_partition->bounds_timestamp(cfg.time)` would return true in between logical prefix truncation (trim-prefix, apply retention) and actual asynchronous physical garbage collection.

Moreover, the way I understand the system is that `trim-prefix` is applied asynchronously. So methods like `bounds_timestamp` can't know whether any data is there or not unless we refactor them to always sync.

Now if local_timequery got a timestamp on the left boundary of the log and also thought that there is data in the cloud log, it returned nullopt. But the cloud storage also failed to find the timestamp because it was fully truncated. The result was "no offset found" but instead local storage should have been authoritative.

My fix was to add:

```
    const bool may_answer_from_cloud
      = may_read_from_cloud()
        && _cloud_storage_partition->bounds_timestamp(cfg.time)
        && cfg.min_offset < kafka::offset_cast(
             _cloud_storage_partition->next_kafka_offset());
```

The `allow_cloud_fallback` was introduced to remove code duplication.